### PR TITLE
INT-4062: Reject Exec. Channel Early Subscription

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/INT-4062

### CODE REVIEW

**Comment by artembilan** [comment]
I wonder if Travis failures are related somehow to your change...

**Comment by garyrussell** [comment]
Looks like it 😄

File: spring-integration-core/src/test/java/org/springframework/integration/channel/PublishSubscribeChannelTests.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ * @since 5.0
+ *
+ */
+public class PublishSubscribeChannelTests {
+
+	@Test
+	public void testEarlySubscribe() {
+		PublishSubscribeChannel channel = new PublishSubscribeChannel(mock(Executor.class));
+		try {
+			channel.subscribe(mock(MessageHandler.class));
```
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/channel/PublishSubscribeChannelTests.java]: Mock? 
You can use pretty simple Lambda:
```
channel.subscribe(m -> {});
```
:smile:

File: spring-integration-core/src/test/java/org/springframework/integration/channel/PublishSubscribeChannelTests.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+			fail("expected Exception");
+		}
+		catch (IllegalStateException e) {
+			assertThat(e.getMessage(), equalTo("When providing an Executor, you cannot subscribe() until the channel "
+					+ "bean is fully initialized by the framework. Do not subscribe in a @Bean definition"));
+		}
+		channel.setBeanFactory(mock(BeanFactory.class));
+		channel.afterPropertiesSet();
+		channel.subscribe(mock(MessageHandler.class));
```
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/channel/PublishSubscribeChannelTests.java]: I think we still should have some assertion in the end to be sure that everything left as expected during normal process.

**Comment by artembilan** [comment]
Why don't we apply similar fix for the `ExecutorChannel`?
Looks like it has the same issue with `dispatcher` overriding in the `onInit()`.

**Comment by garyrussell** [comment]
Moved validation logic to `onInit` and added same for `ExecutorChannel`; squashed.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration/commit/ea5bbe83e5ac0d81701c31de4e3e38c12a50c4d3

**Pull request closed without merging**